### PR TITLE
Add DbWatcher and use PostgreSQL as default

### PR DESCRIPTION
**Pull request by Setogit**
connect to strongloop-internal/scrum-nodeops#899

### CODE REVIEW

**Comment by Setogit** [comment]
@slnode test please

**Comment by Setogit** [comment]
All tests except test-scheduling.js pass. test-scheduling.js should work after listener/watcher side of DbWatcher integration.

**Comment by Setogit** [comment]
@slnode test please

**Comment by Setogit** [comment]
@slnode test please

File: test/helper.js [hunk 1]
```diff
@@ -9,18 +11,49 @@ function createCentralAndTest(title, testFn) {
...
+      var pgCredentials = process.env.POSTGRESQL_USER;
+      if ((process.env.POSTGRESQL_PASSWORD || '').length > 0) {
+        pgCredentials += ':' + process.env.POSTGRESQL_PASSWORD;
+      }
+      // template1 db is guaranteed to exist on all posgres servers
+      var POSTGRESQL_PARAM = fmt('postgres://%s@%s:%d/%s',
+          pgCredentials,
+          process.env.POSTGRESQL_HOST || 'localhost',
+          process.env.POSTGRESQL_PORT || 5432,
+          process.env.POSTGRESQL_DATABASE || 'template1');
```
Comment (rmg) [inline test/helper.js]: You should probably avoid running tests against `template1` if possible. If you modify it, then any other build that runs on the same machine will see those changes and those changes will also be propagated to any new database that is created on that server afterward.

**Comment by Setogit** [comment]
@slnode test please

File: test/helper.js (hunk 1 above)
Comment (Setogit) [inline test/helper.js]: default to 'postgres' insttead

Comment (rmg) [inline test/helper.js]: I was thinking more like `'strong-central-' + process.pid` so that there's little/no potential of cross-contamination of things like triggers and stored procedures or conflicting schemas with the same name.

**Comment by Setogit** [comment]
@slnode test please

**Comment by Setogit** [comment]
@slnode test please

**Comment by Setogit** [comment]
@slnode test please

**Comment by Setogit** [comment]
@slnode test please

**Comment by Setogit** [comment]
All tests but one line passed.

**Comment by Setogit** [comment]
connect to https://github.com/strongloop-internal/scrum-nodeops/issues/899

**Comment by Setogit** [comment]
@slnode test please

File: server/server.js [hunk 2]
```diff
@@ -147,14 +159,76 @@ function start(cb) {
     initEnv,
     reconnectExecutors,
     reconnectGateways,
+    initDbWatcher,
     emitListeningSignal,
   ], done);
 
   function initDatasource(callback) {
     debug('updating database');
+    if (self._meshApp.dataSources.db.connector.name === 'postgresql')
```
Comment (rmg) [inline server/server.js]: No need for this conditional, we can just call `setmaxListeners(0)` every time.

File: server/server.js [hunk 3]
```diff
@@ -147,14 +159,76 @@ function start(cb) {
...
+        // Heavier needs on postgresql for DbWatcher
+        self._meshApp.dataSources.db.setMaxListeners(0);
     self._meshApp.dataSources.db.autoupdate(callback);
   }
 
+  function initDbWatcher(callback) {
+    debug('initializing db watcher');
+    // postgres
+    var connector = self._meshApp.dataSources.db.connector;
+    if (connector.name !== 'postgresql') return;
```
Comment (rmg) [inline server/server.js]: Should be `return callback()`;

File: server/server.js [hunk 4]
```diff
@@ -147,14 +159,76 @@ function start(cb) {
...
+                           params.database);
+    // Setogit TODO: get the list dynamicaly
+    // Issues:
+    //    serverservice is 'Service', not 'PersistedModel'
+    //    group is 'PersistedModel' but doesn't work
+    var tableNames = [];
+    for (var key in connector._models) {
+      var base = connector._models[key].settings.base;
+      if (typeof base === 'string' && base === 'PersistedModel')
+          tableNames.push(key.toLowerCase());
```
Comment (rmg) [inline server/server.js]: `tableNames.push(connector.tableEscaped(connector._models[key]);`

File: server/server.js [hunk 5]
```diff
@@ -110,6 +121,7 @@ function Server(options) {
   // The express app on which the rest of the middleware is mounted.
   this._baseApp = express();
   this._serviceManager = new ServiceManager(this);
+  // TODO: need to initialize this._minkelite
```
Comment (rmg) [inline server/server.js]: This seems familiar. It might be what I was referring to in https://github.com/strongloop-internal/scrum-nodeops/issues/885

File: server/server.js (hunk 2 above)
Comment (Setogit) [inline server/server.js]: removed

File: server/server.js (hunk 3 above)
Comment (Setogit) [inline server/server.js]: added

**Comment by rmg** [comment]
@slnode test please

**Comment by rmg** [comment]
@slnode test please

File: server/server.js (hunk 4 above)
Comment (Setogit) [inline server/server.js]: Now that Broadcast is not required, tableNames are not needed here.  Deleted the Broadcast call and tableNames array all together.

**Comment by Setogit** [comment]
The latest PR of strong-db-watcher supports "lazy broadcast" which does not require Broadcast call.  This PR requires the latest strong-db-watcher.

File: server/server.js [hunk 6]
```diff
@@ -58,14 +61,20 @@ function Server(options) {
...
+        connector: PostgreSQL,
+        url: options.url,
+      };
+      break;
     case 'sqlite3':
       this._dataSourceConfig = {
         connector: SQLite3,
         file: options['mesh.db.filePath'] ||
-          path.join(this._baseDir, 'strong-mesh.db'),
+          path.join(this._baseDir, 'strong-mesh-sqlite3.db'),
```
Comment (kraman) [inline server/server.js]: this change is unnecessary

File: server/server.js [hunk 7]
```diff
@@ -152,9 +163,36 @@ function start(cb) {
...
+    self._dbWatcher = {pgClient: new connector.pg.Client(credential)};
+    self._dbWatcher.pgClient.connect(function(err) {
+      if (err) {
+        if (self._dbWatcher && self._dbWatcher.pgClient)
+            self._dbWatcher.pgClient.end();
+        self._dbWatcher.pgClient = null;
+        return callback(err);
+      }
+      self._dbWatcher.watcher = new DbWatcher(self._dbWatcher.pgClient);
+      self._meshApp.emit('ready', self._dbWatcher.watcher);
```
Comment (kraman) [inline server/server.js]: This can be an api call instead of an event and perhaps pass & call the callback in there as well.

File: server/server.js [hunk 8]
```diff
@@ -265,6 +303,19 @@ function stop(cb) {
     shutdownTasks.push(this._driver.stop.bind(this._driver));
   }
 
+  // Setogit TODO fix up
+  if (this._dbWatcher) {
+    if (this._dbWatcher.watcher) {
+      this._dbWatcher.watcher.close(function(err) {
```
Comment (kraman) [inline server/server.js]: This should be added to `shutdownTasks`

File: test/test-db-driver.js [hunk 9]
```diff
@@ -2,6 +2,7 @@ var Server = require('../server/server');
 var mktmpdir = require('mktmpdir');
 var path = require('path');
 var test = require('tap').test;
+var fmt = require('util').format;
```
Comment (kraman) [inline test/test-db-driver.js]: sort lexically

File: test/helper.js [hunk 10]
```diff
@@ -1,7 +1,9 @@
+var async = require('async');
 var Server = require('../server/server');
 var mktmpdir = require('mktmpdir');
 var test = require('tap').test;
 var url = require('url');
+var fmt = require('util').format;
```
Comment (kraman) [inline test/helper.js]: sort lexically

File: server/server.js [hunk 11]
```diff
@@ -1,11 +1,13 @@
 'use strict';
 
+var DbWatcher = require('strong-db-watcher');
 var EventEmitter = require('events').EventEmitter;
 var ExecutorDriver = require('./drivers/executor');
 var GatewayDriver = require('./gateway');
 var MeshServer = require('strong-mesh-models').meshServer;
 var MinkeLite = require('minkelite');
 var ServiceManager = require('./service-manager');
+var PostgreSQL = require('loopback-connector-postgresql');
```
Comment (kraman) [inline server/server.js]: sort lexically

File: test/test-scheduling.js [hunk 12]
```diff
@@ -135,41 +136,45 @@ createCentralAndTest('register and connect executor',
...
+    //     function ensureNoneExist(model, callback) {
+    //       model.find(function(err, modelInstances) {
+    //         if (err) return callback(err);
+    //         tt.equal(modelInstances.length, 0,
+    //           'No instances of ' + model.modelName + ' should exist');
+    //         callback();
+    //       });
+    //     }
+    //   }
+    // );
```
Comment (kraman) [inline test/test-scheduling.js]: why is this commented out?

File: server/server.js (hunk 6 above)
Comment (Setogit) [inline server/server.js]: reverted

File: server/server.js (hunk 7 above)
Comment (Setogit) [inline server/server.js]: It can be done.  Need to coordinate with watchTable call on the mesh-model side.  Would like to keep the event for this PR if it's not a design flaw, e.g., what if central restarts, etc.

File: server/server.js (hunk 8 above)
Comment (Setogit) [inline server/server.js]: added

File: test/test-db-driver.js (hunk 9 above)
Comment (Setogit) [inline test/test-db-driver.js]: done

File: test/helper.js (hunk 10 above)
Comment (Setogit) [inline test/helper.js]: done

File: server/server.js (hunk 11 above)
Comment (Setogit) [inline server/server.js]: done

File: server/server.js (hunk 7 above)
Comment (kraman) [inline server/server.js]: @Setogit 
* Its an odd way of invoking the setup when calling a method is all its doing. 
* The 'ready' event does not mean anything and is also a very generic term which can be easily confused.
* You are running async functions within the event, but not waiting for them to complete before calling the callback above.

File: test/test-scheduling.js (hunk 12 above)
Comment (Setogit) [inline test/test-scheduling.js]: Because it requires a change in mesh-model client.js serviceDestroy.  Made the change on the mesh side and uncommented.  Then, the test passes.

**Comment by Setogit** [comment]
All comments incorporated except that will keep the event-based mesh/central interaction on dbWatch.  The test in this PR requires a change in mesh-model (in the oher mesh PR).

**Comment by Setogit** [comment]
@slnode test please

**Comment by Setogit** [comment]
connect to https://github.com/strongloop-internal/scrum-nodeops/issues/899

File: server/server.js (hunk 7 above)
Comment (Setogit) [inline server/server.js]: moved the dbWatch initialization to service manager and did it in an API way.

**Comment by Setogit** [comment]
connect to https://github.com/strongloop-internal/scrum-nodeops/issues/899

**Comment by Setogit** [comment]
centralApp.start calls setMaxListeners, but createCentralAndTest needs to call it before that.

**Comment by sam-github** [comment]
@kraman can you link/mention to the loopback-connector-postgresql bug in the column naming, please?

File: server/server.js [hunk 13]
```diff
@@ -182,9 +193,38 @@ function start(cb) {
...
+    self._dbWatcher.pgClient.connect(function(err) {
+      if (err) {
+        if (self._dbWatcher && self._dbWatcher.pgClient)
+            self._dbWatcher.pgClient.end();
+        self._dbWatcher.pgClient = null;
+        return callback(err);
+      }
+
+      debug('initialized db watcher');
+      self._dbWatcher.watcher = new DbWatcher(self._dbWatcher.pgClient);
```
Comment (sam-github) [inline server/server.js]: I think `_dbWatcher.watcher` is confusing, which is the watcher? It would help if the bag of Pg stuff had a better name. Can they both be seperate properties on self? `self._dbClient =  new connector.pg.Client(credential); self._dbWatcher = new DbWatcher()` rather than put into a third thing which then needs a decent name? I had to read this pretty closely to see what watcher was what.

File: server/server.js [hunk 14]
```diff
@@ -303,6 +343,19 @@ function stop(cb) {
...
+    shutdownTasks.push(
+        this._dbWatcher.watcher.close.bind(this._dbWatcher.watcher));
+  }
+
+  if (this._dbWatcher && this._dbWatcher.pgClient) {
+    shutdownTasks.push(function(callback) {
+      this._dbWatcher.pgClient.end();
+      this._dbWatcher.pgClient = null;
+      callback();
+    }.bind(this));
```
Comment (sam-github) [inline server/server.js]: this works, but you usually define `self=this` and use a closure

**Comment by sam-github** [comment]
I found the naming in initDbWatcher made it a bit hard to follow, and have some other small suggestions, but LGTM

**Comment by sam-github** [comment]
LGTM
